dipy/align/streamwarp.py

```diff
@@ -0,0 +1,288 @@
+import warnings
+
+pycpd, have_pycpd, _ = optional_package("pycpd")
+
+if have_pycpd:
+    from pycpd import DeformableRegistration
+
+
+def average_bundle_length(bundle):
+    """Find average Euclidean length of the bundle in mm.
+
+    Parameters
+    ----------
+    bundle : Streamlines
+        Bundle who's average length is to be calculated.
+
+    Returns
+    -------
+    int
+        Average Euclidean length of bundle in mm.
+
+    """
+    metric = AveragePointwiseEuclideanMetric()
+    qb = QuickBundles(threshold=85., metric=metric)
+    clusters = qb.cluster(bundle)
+    centroids = Streamlines(clusters.centroids)
+
+    return length(centroids)[0]
+
+
+def find_missing(lst, cb):
+    """Find unmatched streamline indices in moving bundle.
+
+    Parameters
+    ----------
+    lst : List
+        List of integers containing all the streamlines indices in moving
+        bundle.
+    cb : List
+        List of integers containing streamline indices of the moving bundle
+        that were not matched to any streamline in static bundle.
+
+    Returns
+    -------
+    list
+        List containing unmatched streamlines from moving bundle
+
+    """
+    return [x for x in range(0, len(cb)) if x not in lst]
+
+
+def bundlewarp(static, moving, dist=None, alpha=0.3, beta=20, max_iter=15,
+               affine=True):
+    """Register two bundles using nonlinear method.
+
+    Parameters
+    ----------
+    static : Streamlines
+        Reference/fixed bundle
+
+    moving : Streamlines
+        Target bundle that will be moved/registered to match the static bundle
+
+    dist : float, optional.
+        Precomputed distance matrix (default None)
+
+    alpha : float, optional
+        Represents the trade-off between regularizing the deformation and
+        having points match very closely. Lower value of alpha means high
+        deformations (default 0.3)
+
+    beta : int, optional
+        Represents the strength of the interaction between points
+        Gaussian kernel size (default 20)
+
+    max_iter : int, optional
+        Maximum number of iterations for deformation process in ml-CPD method
+        (default 15)
+
+    affine : boolean, optional
+        If False, use rigid registration as starting point (default True)
+
+    Returns
+    -------
+    deformed_bundle : Streamlines
+        Nonlinearly moved bundle (warped bundle)
+
+    moving_aligned : Streamlines
+        Linearly moved bundle (affinely moved)
+
+    dist : np.ndarray
+        Float array containing distance between moving and static bundle
+
+    matched_pairs : np.ndarray
+        Int array containing streamline correspondences between two bundles
+
+    warp : np.ndarray
+        Nonlinear warp map generated by BundleWarp
+
+    References
+    ----------
+    .. [Chandio2023] Chandio et al. "BundleWarp, streamline-based nonlinear
+            registration of white matter tracts." bioRxiv (2023): 2023-01.
+    """
```

Comment:
here, you need to add 
```python 
if not have_pycpd:
msg = "Missing pycpd dependency,. Please install...." 
   raise ImportError(msg)

Reply:
Should I change it to warning instead? I think tests are failing because of it.

warn(msg)

Reply:
you need to put it back:

```python
if not have_pycpd:
    msg = "Missing pycpd dependency. Please install pycpd package"
    raise ImportError(msg)
```

Test was failing because skipif was missing

Reply:
but I have skipif in tests

Reply:
yes, you have it now, but it was not the case before and you remove that. 

When you added the skipif you should have add this also

Reply:
But I never removed this

`if not have_pycpd:
    msg = "Missing pycpd dependency. Please install pycpd package"
    raise ImportError(msg)`